### PR TITLE
fix Issue 23682 - dip1000 problem with return by ref

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by WalterBright** [comment]
@RazvanN7 it's failing compiling druntime because that gets compiled before the tests/compilable test suite is run. So now I'm doing another reduction job through a thicket of templates. Sigh.

**Comment by WalterBright** [comment]
@RazvanN7 I fixed the problem compiling druntime. But the very next thing it does is try to compile Phobos. And then it fails, and I'm borked again by the order in which the test suite is run.

Every single check attempts to compile Phobos before running any tests.

**Comment by WalterBright** [comment]
This is failing in std.uni, probably the most template heavy piece of code. I can't make much sense of it. It uses auto so much it's really hard to figure out what types are being used. Anyway, being the pragmatic sort, I'm going to put this fix behind a -preview switch. It's likely to break user code, too, that relied on the old behavior. Hopefully this will get things through the test suite.

**Comment by WalterBright** [comment]
oops pushed the wrong button

**Comment by ibuclaw** [comment]
> Every single check attempts to compile Phobos before running any tests.

You can't run any runnable tests without a runtime library to link. You can't run any compilable LINK tests without a runtime library either (though they if any should be standalone programs).

**Comment by WalterBright** [comment]
It should link with the previous version.

File: compiler/test/fail_compilation/retscope6.d [hunk 1]
```diff
@@ -141,6 +141,7 @@ void testarchie()
 
 /* TEST_OUTPUT:
 ---
+fail_compilation/retscope6.d(9015): Error: returning `S9(null).this(r)` escapes a reference to parameter `r`
```
Comment (dkorpel) [inline compiler/test/fail_compilation/retscope6.d:144]: This is false. The code below even says:
```D
auto j = fred(&i); // ok
```

File: druntime/src/rt/monitor_.d [hunk 2]
```diff
@@ -230,7 +230,7 @@ struct Monitor
 
 private:
 
-@property ref shared(Monitor*) monitor(return scope Object h) pure nothrow @nogc
+@property ref shared(Monitor*) monitor(Object h) pure nothrow @nogc
```
Comment (dkorpel) [inline druntime/src/rt/monitor_.d:233]: This signature is correct, the resulting `ref` points into the `Object`, it can be `return scope`

**Review by dkorpel — DISMISSED** [review]
This is the wrong fix.

Fixing a rejects-valid bug shouldn't break druntime and Phobos. 

The function returning by `ref` doesn't mean `return scope` parameters return a reference to the input variable, that would always be a dangling stack pointer. 

The real issue is that assigning a `ref` to a non-`ref` implicitly dereferences a pointer, but this only manifests in the glue layer, so it's invisible to escape.d.

Without redesigning dmd so the dereferencing of `ref` values appears in the front-end AST, here's a possible fix:
https://github.com/dlang/dmd/pull/14871

File: compiler/test/fail_compilation/retscope6.d (hunk 1 above)
Comment (WalterBright) [reply compiler/test/fail_compilation/retscope6.d:144]: hmm, you're right.

**Comment by WalterBright** [comment]
```
****** FAIL release64 std.random
core.exception.AssertError@std/random.d(3957): Assertion failure
```
A random failure in std.random. I wish somehow we could create repeatable tests.

File: druntime/src/rt/monitor_.d (hunk 2 above)
Comment (WalterBright) [reply druntime/src/rt/monitor_.d]: you're right

**Comment by WalterBright** [comment]
Now this new failure on Mac OS:
```
core.exception.AssertError@hospital.d(35): Assertion failure
```
what's going on here?

**Comment by WalterBright** [comment]
And our old friend, the socket failure, reappears:
```
core.exception.AssertError@std/socket.d(805): unittest failure
```

**Comment by WalterBright** [comment]
https://github.com/dlang/phobos/pull/8681

**Comment by WalterBright** [comment]
trying again to see if we avoid heisenbugs

**Comment by WalterBright** [comment]
And the heisenbugs go away! Ready to merge.

**Comment by WalterBright** [comment]
@dkorpel ping

File: compiler/test/fail_compilation/retscope6.d [hunk 3]
```diff
@@ -289,3 +291,33 @@ ref int escape23021() @safe
     // ensure we do not infer return ref
     return infer23021(nonScopePtr); // no error
 }
+
+/******************************/
+
+// https://issues.dlang.org/show_bug.cgi?id=23682
+
+alias VErr = char*;
```
Comment (dkorpel) [inline compiler/test/fail_compilation/retscope6.d:299]: This line is pointless

File: compiler/src/dmd/escape.d [hunk 4]
```diff
@@ -1744,7 +1744,27 @@ void escapeByValue(Expression e, EscapeByResults* er, bool live = false, bool re
                     Parameter p = tf.parameterList[i - j];
                     const stc = tf.parameterStorageClass(null, p);
                     ScopeRef psr = buildScopeRef(stc);
-                    if (psr == ScopeRef.ReturnScope || psr == ScopeRef.Ref_ReturnScope)
```
Comment (dkorpel) [inline compiler/src/dmd/escape.d:1747]: `Ref_ReturnScope` should not be treated different to `ReturnScope`

**Review by dkorpel — APPROVED** [review]
Once again, this fix is overfitted to the single example in the one bugzilla issue, but I can fix this in my PR to solve the issue proper.

File: compiler/src/dmd/escape.d (hunk 4 above)
Comment (WalterBright) [reply compiler/src/dmd/escape.d:1747]: they are different because a ReturnScope can be returned as a Ref

**Comment by WalterBright** [comment]
The trouble is a ref return from a struct constructor is different from a regular ref return. It's an inconsistency in the language, but I don't think there's much we can do about it.

**Pull request merged**